internal/bbr/filter.go

```diff
@@ -0,0 +1,128 @@
+package bbr
+
+import (
+	"math"
+	"time"
+)
+
+// Size of the bottleneck bandwidth filter's window. The paper suggests
+// 6-10, other than that this is arbitrary.
+const btlBwFilterSize = 6
+
+// Filter that estimates the bottleneck bandwidth.
+type btlBwFilter struct {
+	q        queue[int64]
+	Estimate int64
+}
+
+func newBtlBwFilter() btlBwFilter {
+	return btlBwFilter{
+		q: *newQueue[int64](btlBwFilterSize),
+
+		// We set this to something that is non-zero,
+		// but otherwise won't take precedence over any actual
+		// data we receive:
+		Estimate: 1,
+	}
+}
+
+func (f *btlBwFilter) AddSample(deliveryRate int64) {
+	if f.q.Len() == btlBwFilterSize {
+		f.q.Pop()
+	}
+	f.q.Push(deliveryRate)
+	f.Estimate = f.q.Fold(0, max[int64])
+}
+
+// Filter that estimates the round-trip propagation time.
+type rtPropFilter struct {
+	q          queue[rtPropSample]
+	nextSample rtPropSample
+	Estimate   time.Duration
+}
+
+func newRtPropFilter() rtPropFilter {
+	return rtPropFilter{
+		nextSample: rtPropSample{
+			// Set this to a value that will be removed immediately,
+			// and will be lower priority than any other value, so
+			// it will removed from the queue when a new sample is
+			// added.
+			//
+			// XXX: this still could theoretically do the wrong thing
```

Comment:
Would such a failure be transient, or would rtProp be permanently stuck in a bad state?

Reply:
This is stale; originally I had the value as zero but for other reasons I changed it to MaxInt64, so this will be superseded on the first sample in all cases now. Fixed the docs.